drv/cosmo-seq-server/src/main.rs

```diff
@@ -588,6 +597,89 @@ impl ServerImpl {
             set_timer_relative(interval, notifications::TIMER_MASK);
         }
     }
+
+    /// Powers down to A2, if that fails for some reason just
+    /// log an error
+    fn emergency_a2(&mut self, reason: StateChangeReason) {
+        // Power down to A2, updating our internal state.  We can't
+        // handle errors here, so log them and continue.
+        if let Err(e) = self.set_state_impl(PowerState::A2, reason) {
+            ringbuf_entry!(Trace::PowerDownError(e))
+        }
+    }
+
+    fn enable_sequencer_interrupts(&mut self) {
+        self.seq.ier.modify(|m| {
+            m.set_fanfault(true);
+            m.set_thermtrip(true);
+            m.set_smerr_assert(true);
+            m.set_a0mapo(true);
+            m.set_nicmapo(true);
+            m.set_amd_pwrok_fedge(true);
+            m.set_amd_rstn_fedge(true);
+        });
+        let _ = self.sys.gpio_irq_control(
+            notifications::SEQ_IRQ_MASK,
+            sys_api::IrqControl::Enable,
+        );
+    }
+
+    fn disable_sequencer_interrupts(&mut self) {
+        self.seq.ier.modify(|m| {
+            m.set_fanfault(false);
+            m.set_thermtrip(false);
+            m.set_smerr_assert(false);
+            m.set_a0mapo(false);
+            m.set_nicmapo(false);
+            m.set_amd_pwrok_fedge(false);
+            m.set_amd_rstn_fedge(false);
+        });
+        let _ = self.sys.gpio_irq_control(
+            notifications::SEQ_IRQ_MASK,
+            sys_api::IrqControl::Disable,
+        );
+    }
+
+    fn handle_sequencer_interrupt(&mut self) {
+        let ifr = self.seq.ifr.view();
+
+        let state = self.log_state_registers();
+        ringbuf_entry!(Trace::SequencerInterrupt {
+            our_state: self.state,
+            seq_state: state.seq,
+        });
+
+        if ifr.thermtrip {
+            self.seq.ifr.modify(|h| h.set_thermtrip(false));
+            ringbuf_entry!(Trace::Thermtrip);
+            self.set_state_internal(PowerState::A0Thermtrip)
+            // this is a terminal state (for now)
+        }
+        if ifr.amd_pwrok_fedge || ifr.amd_rstn_fedge {
+            let rstn = self.seq.amd_reset_fedges.counts();
+            let pwrokn = self.seq.amd_pwrok_fedges.counts();
+            ringbuf_entry!(Trace::ResetCounts { rstn, pwrokn });
+            // counters are cleared in the A2 -> A0 transition
+            self.set_state_internal(PowerState::A0Reset);
+            // host_sp_comms will be notified of this change and will
+            // call back into this task to reboot the system (going to
+            // A2 then back into A0)
+        }
+
+        if ifr.smerr_assert {
+            // Give up?
+            self.seq.ifr.modify(|h| h.set_smerr_assert(false));
+            self.emergency_a2(StateChangeReason::SmerrAssert);
```

Comment:
Both this and the case below will get us permanently stuck in A2, because `host-sp-comms` will not initiate a power cycle (it only checks for `A0Reset`).  I'm not sure what our policy should be here, any thoughts?

Reply:
I think `smerr_assert` and `a0mapo` are similar to `thermtrip` which looks to also be terminal and won't get restarted (assuming I can read this correctly), I expect this is where we'll want our ereporting to come in to deliver useful information elsewhere.